openfold/train.py

```diff
@@ -0,0 +1,957 @@
+    for _ in range(len(validation_dataloader)):
+        perf = -time.perf_counter()
+        val_batch = next(val_batch_iterator)
+        assert len(val_batch["id"]) == 1
+        id_tuple = val_batch["id"][0]
+        with torch.no_grad():
+            val_batch = map_tensor_tree(
+                fn=lambda t: t.to(device=device),
+                tree=val_batch,
+            )
+            val_outputs = alphafold(val_batch)
+            val_batch = map_tensor_tree(fn=lambda t: t[..., -1], tree=val_batch)
+            val_metrics = compute_validation_metrics(
+                predicted_atom_positions=val_outputs["final_atom_positions"],
+                target_atom_positions=val_batch["all_atom_positions"],
+                atom_mask=val_batch["all_atom_mask"],
+                metrics_names={"lddt_ca"},
+            )
+        perf += time.perf_counter()
+        val_metrics = map_dict_values(fn=lambda t: t.item(), d=val_metrics)
+        val_metrics["val_index"] = id_tuple[1]
+        val_metrics["pdb_chain_id"] = id_tuple[3]
+        val_metrics["duration"] = perf
+        val_metrics_list.append(val_metrics)
+    alphafold.train()
+    return val_metrics_list
+
+
+def training(args: argparse.Namespace) -> None:
+    if args.distributed:
+        torch.distributed.init_process_group(backend="nccl", init_method="env://")
+
+    if torch.distributed.is_initialized():
+        # Assuming distributed training:
+        assert args.distributed is True
+        # https://pytorch.org/docs/stable/elastic/run.html#environment-variables
+        rank = int(os.environ["RANK"])
+        world_size = int(os.environ["WORLD_SIZE"])
+        local_world_size = int(os.environ["LOCAL_WORLD_SIZE"])
+        local_rank = int(os.environ["LOCAL_RANK"])
+        assert world_size % local_world_size == 0
+        num_nodes = world_size // local_world_size
+        main_rank = 0
+        is_main_process = bool(rank == main_rank)
+        process_name = f"dist_process_rank{rank}"
+        device = torch.device(f"cuda:{local_rank}")
+        global_batch_size = args.local_batch_size * world_size
+        if is_main_process:
+            print(f"initialized distributed training: WORLD_SIZE={world_size}")
+    else:
+        # Assuming single GPU training:
+        print("single GPU training")
+        assert args.distributed is False
+        rank = None
+        world_size = None
+        local_world_size = None
+        local_rank = None
+        num_nodes = None
+        main_rank = None
+        is_main_process = True
+        process_name = "single_process"
+        device = torch.device("cuda:0")
+        global_batch_size = args.local_batch_size
+
+    # Create output directory:
+    args.training_dirpath.mkdir(parents=True, exist_ok=True)
+
+    # MLPerf logging setup:
+    mllog_datestamp = os.environ.get("DATESTAMP", "yymmddHHMMSSfffffffff")
+    mlperf_instance = "0"  # TODO: set this value correctly for "Throughput" benchmark
```

Comment:
There is a TODO here about setting the mlperf instance number. I guess this could be needed for the log file generation unless they are managed separately.

Reply:
I don't want to block approval on this, so I'll just go ahead and approve, but it'd be good to have a response on whether anything is needed.

Reply:
Resolved in this [commit](https://github.com/ar-nowaczynski/mlcommons-hpc/commit/b7ddd769409e08bc49d853f17a694cf96ba737c4).

When launching throughput-oriented benchmark, training script expects integer value for `mlperf_throughput_num_instances`, which is later used to determine `mlperf_instance`.

Reference code is not expected to fully support throughput-oriented benchmark workflow. This change alone is not sufficient to enable it. But it is good default behavior in case of unknown.